File: ios/CallPilot/Protocol/CallContentModels.swift
import Foundation

enum CallDirection: String, Codable, CaseIterable {
    case inbound = "INBOUND"
    case outbound = "OUTBOUND"
}

struct CallRecordStatus: RawRepresentable, Codable, Equatable, Hashable {
    let rawValue: String

    static let completed = Self(rawValue: "COMPLETED")
    static let notConnected = Self(rawValue: "NOT_CONNECTED")
    static let failed = Self(rawValue: "FAILED")
    static let unknown = Self(rawValue: "UNKNOWN")

    init(rawValue: String) { self.rawValue = rawValue }

    init(from decoder: Decoder) throws {
        let value = try decoder.singleValueContainer().decode(String.self)
        guard value.wholeMatch(of: /^[A-Z][A-Z0-9_]{2,63}$/) != nil else {
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath, debugDescription: "Invalid product call status")
            )
        }
        rawValue = value
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

enum CallSource: String, Codable, CaseIterable {
    case agent = "AGENT"
    case remoteHandset = "REMOTE_HANDSET"
    case unknown = "UNKNOWN"
}

enum CallSummaryState: String, Codable, CaseIterable {
    case pending = "PENDING"
    case ready = "READY"
    case failed = "FAILED"
    case unavailable = "UNAVAILABLE"
}

enum CallTriageOutcome: String, Codable, CaseIterable {
    case aiHandled = "AI_HANDLED"
    case rejected = "REJECTED"
    case transferred = "TRANSFERRED"
    case unknown = "UNKNOWN"
}

struct CallRecordItem: Codable, Equatable, Hashable, Identifiable {
    let callId: String
    let revision: String
    let direction: CallDirection
    let address: String?
    let startedAt: Int64
    let endedAt: Int64?
    let durationMs: Int64?
    let status: CallRecordStatus
    let answered: Bool
    let source: CallSource
    let summaryState: CallSummaryState
    let summaryPreview: String?
    let hasTranscript: Bool
    let triageOutcome: CallTriageOutcome?

    var id: String { callId }

    private static var callIdRE: Regex<Substring> { /^call_[A-Za-z0-9_-]{12,80}$/ }
    private static var revisionRE: Regex<Substring> { /^revision_[A-Za-z0-9_-]{12,80}$/ }

    init(
        callId: String,
        revision: String,
        direction: CallDirection,
        address: String?,
        startedAt: Int64,
        endedAt: Int64?,
        durationMs: Int64?,
        status: CallRecordStatus,
        answered: Bool,
        source: CallSource,
        summaryState: CallSummaryState,
        summaryPreview: String?,
        hasTranscript: Bool,
        triageOutcome: CallTriageOutcome?
    ) {
        self.callId = callId
        self.revision = revision
        self.direction = direction
        self.address = address
        self.startedAt = startedAt
        self.endedAt = endedAt
        self.durationMs = durationMs
        self.status = status
        self.answered = answered
        self.source = source
        self.summaryState = summaryState
        self.summaryPreview = summaryPreview
        self.hasTranscript = hasTranscript
        self.triageOutcome = triageOutcome
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let callId = try values.decode(String.self, forKey: .callId)
        let revision = try values.decode(String.self, forKey: .revision)
        let direction = try values.decode(CallDirection.self, forKey: .direction)
        let address = try values.decodeIfPresent(String.self, forKey: .address)
        let startedAt = try values.decode(Int64.self, forKey: .startedAt)
        let endedAt = try values.decodeIfPresent(Int64.self, forKey: .endedAt)
        let durationMs = try values.decodeIfPresent(Int64.self, forKey: .durationMs)
        let status = try values.decode(CallRecordStatus.self, forKey: .status)
        let answered = try values.decode(Bool.self, forKey: .answered)
        let source = try values.decode(CallSource.self, forKey: .source)
        let summaryState = try values.decode(CallSummaryState.self, forKey: .summaryState)
        let summaryPreview = try values.decodeIfPresent(String.self, forKey: .summaryPreview)
        let hasTranscript = try values.decode(Bool.self, forKey: .hasTranscript)
        let triageOutcome = try values.decodeIfPresent(CallTriageOutcome.self, forKey: .triageOutcome)

        guard callId.wholeMatch(of: Self.callIdRE) != nil,
              revision.wholeMatch(of: Self.revisionRE) != nil,
              startedAt >= 0,
              endedAt == nil || endedAt! >= startedAt,
              durationMs == nil || durationMs! >= 0,
              (endedAt == nil) == (durationMs == nil) else {
            throw DecodingError.dataCorruptedError(
                forKey: .callId,
                in: values,
                debugDescription: "Invalid content-sync call record"
            )
        }

        self.init(
            callId: callId, revision: revision, direction: direction, address: address,
            startedAt: startedAt, endedAt: endedAt, durationMs: durationMs, status: status,
            answered: answered, source: source, summaryState: summaryState,
            summaryPreview: summaryPreview, hasTranscript: hasTranscript,
            triageOutcome: triageOutcome
        )
    }
}

struct CallRecordsPage: Codable, Equatable {
    let v: Int
    let items: [CallRecordItem]
    let nextCursor: String?
    let hasMore: Bool
    let collectionRevision: String
    let oldestAvailableAt: Int64?

    init(
        v: Int,
        items: [CallRecordItem],
        nextCursor: String?,
        hasMore: Bool,
        collectionRevision: String,
        oldestAvailableAt: Int64?
    ) {
        self.v = v
        self.items = items
        self.nextCursor = nextCursor
        self.hasMore = hasMore
        self.collectionRevision = collectionRevision
        self.oldestAvailableAt = oldestAvailableAt
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let v = try values.decode(Int.self, forKey: .v)
        let items = try values.decode([CallRecordItem].self, forKey: .items)
        let nextCursor = try values.decodeIfPresent(String.self, forKey: .nextCursor)
        let hasMore = try values.decode(Bool.self, forKey: .hasMore)
        let revision = try values.decode(String.self, forKey: .collectionRevision)
        let oldest = try values.decodeIfPresent(Int64.self, forKey: .oldestAvailableAt)
        guard Self.validPage(
            v: v, ids: items.map(\.callId), nextCursor: nextCursor,
            hasMore: hasMore, revision: revision, oldest: oldest, itemCount: items.count
        ) else {
            throw DecodingError.dataCorruptedError(
                forKey: .v, in: values, debugDescription: "Invalid call records page"
            )
        }
        self.init(
            v: v, items: items, nextCursor: nextCursor, hasMore: hasMore,
            collectionRevision: revision, oldestAvailableAt: oldest
        )
    }

    private static func validPage(
        v: Int, ids: [String], nextCursor: String?, hasMore: Bool,
        revision: String, oldest: Int64?, itemCount: Int
    ) -> Bool {
        v == 1 && itemCount <= 100 && Set(ids).count == ids.count
            && hasMore == (nextCursor != nil)
            && ContentWireValidation.validCursor(nextCursor)
            && ContentWireValidation.validRevision(revision)
            && (oldest == nil || oldest! >= 0)
    }
}

struct CallSummary: Codable, Equatable, Hashable {
    let ok: Bool
    let text: String?
    let callerIdentity: String?
    let intent: String?
    let urgency: String?
    let callbackNeeded: Bool?
    let errorCode: String?
    let resultSource: String?
    let resultVerification: String?
}

struct CallRecordDetail: Codable, Equatable, Hashable {
    let v: Int
    let record: CallRecordItem
    let summary: CallSummary?
    let timelineRevision: String

    init(v: Int, record: CallRecordItem, summary: CallSummary?, timelineRevision: String) {
        self.v = v
        self.record = record
        self.summary = summary
        self.timelineRevision = timelineRevision
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let v = try values.decode(Int.self, forKey: .v)
        let record = try values.decode(CallRecordItem.self, forKey: .record)
        let summary = try values.decodeIfPresent(CallSummary.self, forKey: .summary)
        let timelineRevision = try values.decode(String.self, forKey: .timelineRevision)
        let mustBeNil = record.summaryState == .pending || record.summaryState == .unavailable
        guard v == 1,
              mustBeNil == (summary == nil),
              ContentWireValidation.validRevision(timelineRevision) else {
            throw DecodingError.dataCorruptedError(
                forKey: .summary, in: values, debugDescription: "Invalid call record detail"
            )
        }
        self.init(v: v, record: record, summary: summary, timelineRevision: timelineRevision)
    }
}

enum CallTimelineKind: String, Codable {
    case transcript
    case result
    case triage
    case takeover
    case unknown
}

enum TimelineRole: String, Codable { case agent = "AGENT"; case caller = "CALLER" }
enum TriageCategory: String, Codable { case marketing = "MARKETING"; case personal = "PERSONAL"; case needsOwner = "NEEDS_OWNER"; case unknown = "UNKNOWN" }
enum TriageAction: String, Codable { case clarify = "CLARIFY"; case continueAI = "CONTINUE_AI"; case reject = "REJECT"; case transfer = "TRANSFER" }
enum TakeoverState: String, Codable { case requested = "REQUESTED"; case committed = "COMMITTED"; case ownerHangup = "OWNER_HANGUP"; case failed = "FAILED" }

struct TimelineTranscript: Codable, Equatable, Hashable {
    let timelineItemId: String
    let occurredAt: Int64
    let type: String
    let role: TimelineRole
    let text: String
}

struct TimelineResult: Codable, Equatable, Hashable {
    let timelineItemId: String
    let occurredAt: Int64
    let type: String
    let status: CallRecordStatus
    let summary: String?
}

struct TimelineTriage: Codable, Equatable, Hashable {
    let timelineItemId: String
    let occurredAt: Int64
    let type: String
    let category: TriageCategory
    let action: TriageAction
    let confidence: Double
    let reasonCode: String
}

struct TimelineTakeover: Codable, Equatable, Hashable {
    let timelineItemId: String
    let occurredAt: Int64
    let type: String
    let state: TakeoverState
    let reasonCode: String?
}

struct TimelineUnknown: Codable, Equatable, Hashable {
    let timelineItemId: String
    let occurredAt: Int64
    let type: String
}

enum CallTimelineItem: Codable, Equatable, Hashable, Identifiable {
    case transcript(TimelineTranscript)
    case result(TimelineResult)
    case triage(TimelineTriage)
    case takeover(TimelineTakeover)
    case unknown(TimelineUnknown)

    var id: String { base.timelineItemId }
    var occurredAt: Int64 { base.occurredAt }
    var kind: CallTimelineKind {
        switch self {
        case .transcript: .transcript
        case .result: .result
        case .triage: .triage
        case .takeover: .takeover
        case .unknown: .unknown
        }
    }

    private var base: TimelineUnknown {
        switch self {
        case .transcript(let value): .init(timelineItemId: value.timelineItemId, occurredAt: value.occurredAt, type: value.type)
        case .result(let value): .init(timelineItemId: value.timelineItemId, occurredAt: value.occurredAt, type: value.type)
        case .triage(let value): .init(timelineItemId: value.timelineItemId, occurredAt: value.occurredAt, type: value.type)
        case .takeover(let value): .init(timelineItemId: value.timelineItemId, occurredAt: value.occurredAt, type: value.type)
        case .unknown(let value): value
        }
    }

    init(from decoder: Decoder) throws {
        let base = try TimelineUnknown(from: decoder)
        guard ContentWireValidation.validTimelineItemId(base.timelineItemId),
              base.occurredAt >= 0 else {
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath, debugDescription: "Invalid timeline item")
            )
        }
        switch base.type {
        case "TRANSCRIPT": self = .transcript(try TimelineTranscript(from: decoder))
        case "RESULT": self = .result(try TimelineResult(from: decoder))
        case "TRIAGE":
            let item = try TimelineTriage(from: decoder)
            guard (0...1).contains(item.confidence) else {
                throw DecodingError.dataCorrupted(
                    .init(codingPath: decoder.codingPath, debugDescription: "Invalid triage confidence")
                )
            }
            self = .triage(item)
        case "TAKEOVER": self = .takeover(try TimelineTakeover(from: decoder))
        default:
            guard base.type.wholeMatch(of: /^[A-Z][A-Z0-9_]{2,63}$/) != nil else {
                throw DecodingError.dataCorrupted(
                    .init(codingPath: decoder.codingPath, debugDescription: "Invalid future timeline type")
                )
            }
            self = .unknown(base)
        }
    }

    func encode(to encoder: Encoder) throws {
        switch self {
        case .transcript(let value): try value.encode(to: encoder)
        case .result(let value): try value.encode(to: encoder)
        case .triage(let value): try value.encode(to: encoder)
        case .takeover(let value): try value.encode(to: encoder)
        case .unknown(let value): try value.encode(to: encoder)
        }
    }
}

struct CallTimelinePage: Codable, Equatable {
    let v: Int
    let items: [CallTimelineItem]
    let nextCursor: String?
    let hasMore: Bool
    let collectionRevision: String
    let oldestAvailableAt: Int64?

    var visibleItems: [CallTimelineItem] { items.filter { $0.kind != .unknown } }

    init(
        v: Int,
        items: [CallTimelineItem],
        nextCursor: String?,
        hasMore: Bool,
        collectionRevision: String,
        oldestAvailableAt: Int64?
    ) {
        self.v = v
        self.items = items
        self.nextCursor = nextCursor
        self.hasMore = hasMore
        self.collectionRevision = collectionRevision
        self.oldestAvailableAt = oldestAvailableAt
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let v = try values.decode(Int.self, forKey: .v)
        let items = try values.decode([CallTimelineItem].self, forKey: .items)
        let cursor = try values.decodeIfPresent(String.self, forKey: .nextCursor)
        let hasMore = try values.decode(Bool.self, forKey: .hasMore)
        let revision = try values.decode(String.self, forKey: .collectionRevision)
        let oldest = try values.decodeIfPresent(Int64.self, forKey: .oldestAvailableAt)
        guard v == 1, items.count <= 100,
              Set(items.map(\.id)).count == items.count,
              hasMore == (cursor != nil),
              ContentWireValidation.validCursor(cursor),
              ContentWireValidation.validRevision(revision),
              oldest == nil || oldest! >= 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .v, in: values, debugDescription: "Invalid call timeline page"
            )
        }
        self.init(
            v: v, items: items, nextCursor: cursor, hasMore: hasMore,
            collectionRevision: revision, oldestAvailableAt: oldest
        )
    }
}

enum ContentWireValidation {
    static func validCursor(_ value: String?) -> Bool {
        guard let value else { return true }
        return value.count <= 2_048
            && value.wholeMatch(of: /^cursor_[A-Za-z0-9_-]+$/) != nil
    }

    static func validRevision(_ value: String) -> Bool {
        value.wholeMatch(of: /^revision_[A-Za-z0-9_-]{12,80}$/) != nil
    }

    static func validTimelineItemId(_ value: String) -> Bool {
        value.wholeMatch(of: /^item_[A-Za-z0-9_-]{12,80}$/) != nil
    }
}

@MainActor
protocol CallRecordContentClient: AnyObject {
    func listCallRecords(limit: Int, cursor: String?) async throws -> CallRecordsPage
    func getCallRecord(callId: String) async throws -> CallRecordDetail
    func listCallTimeline(callId: String, limit: Int, cursor: String?) async throws -> CallTimelinePage
}

File: android/app/src/main/kotlin/ai/bondings/callpilot/protocol/CallContentModels.kt
package ai.bondings.callpilot.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

@Serializable
enum class CallDirection { INBOUND, OUTBOUND }

@JvmInline
@Serializable
value class CallRecordStatus(val rawValue: String) {
    companion object {
        val COMPLETED = CallRecordStatus("COMPLETED")
        val NOT_CONNECTED = CallRecordStatus("NOT_CONNECTED")
        val FAILED = CallRecordStatus("FAILED")
        val UNKNOWN = CallRecordStatus("UNKNOWN")
    }
}

@Serializable
enum class CallSource { AGENT, REMOTE_HANDSET, UNKNOWN }

@Serializable
enum class CallSummaryState { PENDING, READY, FAILED, UNAVAILABLE }

@Serializable
enum class CallTriageOutcome { AI_HANDLED, REJECTED, TRANSFERRED, UNKNOWN }

@Serializable
data class CallRecordItem(
    val callId: String,
    val revision: String,
    val direction: CallDirection,
    val address: String?,
    val startedAt: Long,
    val endedAt: Long?,
    val durationMs: Long?,
    val status: CallRecordStatus,
    val answered: Boolean,
    val source: CallSource,
    val summaryState: CallSummaryState,
    val summaryPreview: String?,
    val hasTranscript: Boolean,
    val triageOutcome: CallTriageOutcome?,
) {
    fun requireValid() {
        requireContentContract(ContentWireValidation.validCallId(callId))
        requireContentContract(ContentWireValidation.validRevision(revision))
        requireContentContract(ContentWireValidation.validProductCode(status.rawValue))
        requireContentContract(startedAt >= 0)
        requireContentContract(endedAt == null || endedAt >= startedAt)
        requireContentContract(durationMs == null || durationMs >= 0)
        requireContentContract((endedAt == null) == (durationMs == null))
    }
}

@Serializable
data class CallRecordsPage(
    val v: Int,
    val items: List<CallRecordItem>,
    val nextCursor: String?,
    val hasMore: Boolean,
    val collectionRevision: String,
    val oldestAvailableAt: Long?,
) {
    fun requireValid() {
        requireContentContract(v == 1 && items.size <= 100)
        items.forEach(CallRecordItem::requireValid)
        requireContentContract(items.map(CallRecordItem::callId).toSet().size == items.size)
        requireContentContract(hasMore == (nextCursor != null))
        requireContentContract(ContentWireValidation.validCursor(nextCursor))
        requireContentContract(ContentWireValidation.validRevision(collectionRevision))
        requireContentContract(oldestAvailableAt == null || oldestAvailableAt >= 0)
    }

    companion object {
        fun decode(json: Json, text: String): CallRecordsPage = decodeContent(json, text) { it.requireValid() }
    }
}

@Serializable
data class CallSummary(
    val ok: Boolean,
    val text: String?,
    val callerIdentity: String?,
    val intent: String?,
    val urgency: String?,
    val callbackNeeded: Boolean?,
    val errorCode: String?,
    val resultSource: String?,
    val resultVerification: String?,
)

@Serializable
data class CallRecordDetail(
    val v: Int,
    val record: CallRecordItem,
    val summary: CallSummary?,
    val timelineRevision: String,
) {
    fun requireValid() {
        requireContentContract(v == 1)
        record.requireValid()
        val summaryMustBeNil = record.summaryState in setOf(
            CallSummaryState.PENDING,
            CallSummaryState.UNAVAILABLE,
        )
        requireContentContract(summaryMustBeNil == (summary == null))
        requireContentContract(ContentWireValidation.validRevision(timelineRevision))
    }

    companion object {
        fun decode(json: Json, text: String): CallRecordDetail = decodeContent(json, text) { it.requireValid() }
    }
}

enum class CallTimelineKind { TRANSCRIPT, RESULT, TRIAGE, TAKEOVER, UNKNOWN }

@Serializable
enum class TimelineRole { AGENT, CALLER }

@Serializable
enum class TriageCategory { MARKETING, PERSONAL, NEEDS_OWNER, UNKNOWN }

@Serializable
enum class TriageAction { CLARIFY, CONTINUE_AI, REJECT, TRANSFER }

@Serializable
enum class TakeoverState { REQUESTED, COMMITTED, OWNER_HANGUP, FAILED }

@Serializable
data class CallTimelineItem(
    val timelineItemId: String,
    val occurredAt: Long,
    val type: String,
    @SerialName("role") private val roleValue: JsonElement? = null,
    @SerialName("text") private val textValue: JsonElement? = null,
    @SerialName("status") private val statusValue: JsonElement? = null,
    @SerialName("summary") private val summaryValue: JsonElement? = null,
    @SerialName("category") private val categoryValue: JsonElement? = null,
    @SerialName("action") private val actionValue: JsonElement? = null,
    @SerialName("confidence") private val confidenceValue: JsonElement? = null,
    @SerialName("reasonCode") private val reasonCodeValue: JsonElement? = null,
    @SerialName("state") private val stateValue: JsonElement? = null,
) {
    val kind: CallTimelineKind
        get() = when (type) {
            "TRANSCRIPT" -> CallTimelineKind.TRANSCRIPT
            "RESULT" -> CallTimelineKind.RESULT
            "TRIAGE" -> CallTimelineKind.TRIAGE
            "TAKEOVER" -> CallTimelineKind.TAKEOVER
            else -> CallTimelineKind.UNKNOWN
        }

    val timelineRole: TimelineRole? get() = TimelineRole.entries.firstOrNull { it.name == roleValue.stringOrNull() }
    val text: String? get() = textValue.stringOrNull()
    val recordStatus: CallRecordStatus? get() = statusValue.stringOrNull()?.let(::CallRecordStatus)
    val summary: String? get() = summaryValue.stringOrNull()
    val triageCategory: TriageCategory? get() = TriageCategory.entries.firstOrNull { it.name == categoryValue.stringOrNull() }
    val triageAction: TriageAction? get() = TriageAction.entries.firstOrNull { it.name == actionValue.stringOrNull() }
    val confidence: Double? get() = (confidenceValue as? JsonPrimitive)?.takeIf { !it.isString }?.doubleOrNull
    val reasonCode: String? get() = reasonCodeValue.stringOrNull()
    val takeoverState: TakeoverState? get() = TakeoverState.entries.firstOrNull { it.name == stateValue.stringOrNull() }

    fun requireValid() {
        requireContentContract(ContentWireValidation.validTimelineItemId(timelineItemId))
        requireContentContract(occurredAt >= 0)
        requireContentContract(ContentWireValidation.validProductCode(type))
        when (kind) {
            CallTimelineKind.TRANSCRIPT -> requireContentContract(timelineRole != null && text != null)
            CallTimelineKind.RESULT -> requireContentContract(
                recordStatus != null && ContentWireValidation.validProductCode(recordStatus!!.rawValue),
            )
            CallTimelineKind.TRIAGE -> {
                val parsedConfidence = confidence
                val parsedReasonCode = reasonCode
                requireContentContract(
                    triageCategory != null && triageAction != null && parsedConfidence != null &&
                        parsedConfidence in 0.0..1.0 && parsedReasonCode != null &&
                        ContentWireValidation.validProductCode(parsedReasonCode),
                )
            }
            CallTimelineKind.TAKEOVER -> {
                val parsedReasonCode = reasonCode
                requireContentContract(
                    takeoverState != null &&
                        (parsedReasonCode == null || ContentWireValidation.validProductCode(parsedReasonCode)),
                )
            }
            CallTimelineKind.UNKNOWN -> Unit
        }
    }
}

private fun JsonElement?.stringOrNull(): String? =
    (this as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content

@Serializable
data class CallTimelinePage(
    val v: Int,
    val items: List<CallTimelineItem>,
    val nextCursor: String?,
    val hasMore: Boolean,
    val collectionRevision: String,
    val oldestAvailableAt: Long?,
) {
    val visibleItems: List<CallTimelineItem> get() = items.filter { it.kind != CallTimelineKind.UNKNOWN }

    fun requireValid() {
        requireContentContract(v == 1 && items.size <= 100)
        items.forEach(CallTimelineItem::requireValid)
        requireContentContract(items.map(CallTimelineItem::timelineItemId).toSet().size == items.size)
        requireContentContract(hasMore == (nextCursor != null))
        requireContentContract(ContentWireValidation.validCursor(nextCursor))
        requireContentContract(ContentWireValidation.validRevision(collectionRevision))
        requireContentContract(oldestAvailableAt == null || oldestAvailableAt >= 0)
    }

    companion object {
        fun decode(json: Json, text: String): CallTimelinePage = decodeContent(json, text) { it.requireValid() }
    }
}

interface CallRecordContentClient {
    suspend fun listCallRecords(limit: Int, cursor: String?): CallRecordsPage
    suspend fun getCallRecord(callId: String): CallRecordDetail
    suspend fun listCallTimeline(callId: String, limit: Int, cursor: String?): CallTimelinePage
}

private inline fun <reified T> decodeContent(json: Json, text: String, validate: (T) -> Unit): T = try {
    json.decodeFromString<T>(text).also(validate)
} catch (error: ContentContractException) {
    throw error
} catch (error: SerializationException) {
    throw ContentContractException(error)
} catch (error: IllegalArgumentException) {
    throw ContentContractException(error)
}

private fun requireContentContract(condition: Boolean) {
    if (!condition) throw ContentContractException()
}
